File: code/packages/python/process-manager/src/process_manager/process_manager.py
"""Process Manager — Core process lifecycle management.

This module implements the four fundamental Unix process operations:

    fork()  — Clone a running process.
    exec()  — Replace a process's program with a new one.
    wait()  — Wait for a child process to exit and retrieve its exit code.
    kill()  — Send a signal to a process.

Together with exit(), these operations form the complete process lifecycle.
Every program you have ever run on a Unix system was created by fork+exec
and cleaned up by exit+wait.

How Your Shell Works
====================

When you type "ls" in a terminal, this is what happens:

    1. The shell (e.g., bash) calls fork().
       Now there are TWO copies of the shell running.

    2. The child copy calls exec("ls").
       The child is no longer a shell — it is now the "ls" program.

    3. "ls" runs, prints the directory listing, and calls exit(0).
       The child becomes a zombie (state = ZOMBIE).

    4. The parent shell calls wait().
       It receives the child's exit code (0 = success).
       The zombie is reaped (PCB removed from the process table).

    5. The parent shell prints the next prompt. Ready for the next command.

This fork+exec pattern is used everywhere in Unix:
    - Web servers fork() to handle each client connection.
    - Daemons fork() twice to detach from the terminal.
    - init (PID 1) fork+execs every system service at boot.
    - Shells fork+exec every command you type.

The PID 0 Convention
====================

PID 0 is special in our system. It acts as the "init" process — the ancestor
of all other processes. When a process exits without its parent waiting,
its children are "reparented" to PID 0. This prevents orphaned zombies from
accumulating, because PID 0 (init) periodically reaps them.

In real Unix systems, the idle process is PID 0 and init is PID 1, but for
our educational implementation, we combine them into PID 0.
"""

from __future__ import annotations

from process_manager.pcb import ProcessControlBlock, ProcessState
from process_manager.signals import Signal, SignalManager


class ProcessManager:
    """Manages the complete process lifecycle.

    The ProcessManager is the kernel's process subsystem. It owns the process
    table (a dictionary mapping PIDs to PCBs) and provides the system calls
    that user programs invoke to create, manage, and terminate processes.

    Attributes:
        _processes: The process table. Maps PID -> ProcessControlBlock.
        _next_pid: The next PID to assign. Incremented after each allocation.
        _signal_manager: Handles signal delivery for all processes.

    Example:
        >>> pm = ProcessManager()
        >>> init = pm.create_process(name="init", priority=0)
        >>> init.pid
        0
        >>> child_pid, child_ret = pm.fork(init.pid)
        >>> child_ret  # child sees 0 from fork
        0
    """

    def __init__(self) -> None:
        """Initialize an empty process manager.

        The process table starts empty. The first call to create_process()
        will allocate PID 0 (the init process).
        """
        self._processes: dict[int, ProcessControlBlock] = {}
        self._next_pid: int = 0
        self._signal_manager = SignalManager()

    # =========================================================================
    # Process Creation
    # =========================================================================

    def create_process(
        self,
        name: str = "",
        parent_pid: int = -1,
        priority: int = 20,
        memory_base: int = 0,
        memory_size: int = 0,
    ) -> ProcessControlBlock:
        """Create a new process with a unique PID.

        This is the low-level process creation mechanism. It allocates a fresh
        PCB, assigns a unique PID, and adds it to the process table. Higher-
        level operations like fork() use this internally.

        Args:
            name: Human-readable name for debugging (e.g., "shell", "ls").
            parent_pid: PID of the parent process. -1 for the root process.
            priority: Scheduling priority, 0 (highest) to 39 (lowest).
                      Default is 20 (normal user process).
            memory_base: Starting address of the process's memory region.
            memory_size: Size of the process's memory region in bytes.

        Returns:
            The newly created ProcessControlBlock.

        Example:
            >>> pm = ProcessManager()
            >>> pcb = pm.create_process(name="init", priority=0)
            >>> pcb.pid
            0
            >>> pcb.state
            <ProcessState.READY: 0>
        """
        pid = self._next_pid
        self._next_pid += 1

        pcb = ProcessControlBlock(
            pid=pid,
            name=name,
            state=ProcessState.READY,
            parent_pid=parent_pid,
            priority=priority,
            memory_base=memory_base,
            memory_size=memory_size,
        )

        self._processes[pid] = pcb

        # If this process has a parent, add it to the parent's children list.
        if parent_pid >= 0 and parent_pid in self._processes:
            self._processes[parent_pid].children.append(pid)

        return pcb

    # =========================================================================
    # fork() — Clone a Process
    # =========================================================================
    #
    # fork() is the most unusual system call in computing. It creates a new
    # process that is an EXACT COPY of the calling process. Both processes
    # resume at the same point in the code, but they receive different return
    # values:
    #
    #   - The parent receives the child's PID (a positive integer).
    #   - The child receives 0.
    #
    # This is how the program knows which copy it is:
    #
    #   pid = fork()
    #   if pid == 0:
    #       print("I am the child!")
    #   else:
    #       print(f"I am the parent. My child is PID {pid}.")

    def fork(self, parent_pid: int) -> tuple[int, int]:
        """Fork: create a child process as a copy of the parent.

        The child gets:
            - A new, unique PID.
            - A copy of the parent's registers (including PC).
            - The same priority as the parent.
            - The same memory base and size as the parent.
            - The same signal handlers as the parent.
            - An empty children list (no grandchildren).
            - No pending signals (clean slate).
            - cpu_time reset to 0.

        The parent gets:
            - The child's PID added to its children list.

        Returns:
            A tuple (child_pid, 0) where:
                - child_pid is the PID assigned to the new child process.
                  The parent would see this as the return value of fork().
                - 0 is what the child would see as the return value of fork().

            In a real kernel, fork() returns child_pid to the parent and 0 to
            the child by writing different values into their respective a0
            registers. Here, we return both values as a tuple so the caller
            can simulate this behavior.

        Raises:
            ValueError: If parent_pid does not exist in the process table.

        Example:
            >>> pm = ProcessManager()
            >>> parent = pm.create_process(name="shell")
            >>> parent.registers[10] = 42  # some state in a0
            >>> child_pid, child_ret = pm.fork(parent.pid)
            >>> child_ret
            0
            >>> child = pm.get_process(child_pid)
            >>> child.parent_pid == parent.pid
            True
        """
        parent = self._processes.get(parent_pid)
        if parent is None:
            msg = f"No process with PID {parent_pid}"
            raise ValueError(msg)

        # Allocate a new PID for the child.
        child_pid = self._next_pid
        self._next_pid += 1

        # Create the child PCB as a copy of the parent.
        # We copy each field explicitly to make it clear what is copied,
        # what is shared, and what is reset.
        child = ProcessControlBlock(
            pid=child_pid,
            name=parent.name,  # Same name (will be changed by exec).
            state=ProcessState.READY,  # Child starts ready to run.
            # --- Copied from parent ---
            registers=list(parent.registers),  # Deep copy of registers.
            pc=parent.pc,  # Same program counter.
            sp=parent.sp,  # Same stack pointer.
            memory_base=parent.memory_base,  # Same memory region.
            memory_size=parent.memory_size,
            priority=parent.priority,  # Inherit priority.
            signal_handlers=dict(parent.signal_handlers),  # Copy handlers.
            # --- Reset for child ---
            parent_pid=parent_pid,  # My parent is the caller.
            children=[],  # I have no children yet.
            pending_signals=[],  # No pending signals.
            signal_mask=set(parent.signal_mask),  # Copy mask.
            cpu_time=0,  # Fresh CPU time counter.
            exit_code=0,
        )

        # Add child to the process table.
        self._processes[child_pid] = child

        # Update parent's children list.
        parent.children.append(child_pid)

        # Return (child_pid_for_parent, 0_for_child).
        # In a real kernel, we would write child_pid into parent's a0
        # register and 0 into child's a0 register. Here, we return both
        # values so the test/caller can verify.
        return (child_pid, 0)

    # =========================================================================
    # exec() — Replace Process Image
    # =========================================================================
    #
    # exec() throws away the current program and loads a new one. It is like
    # erasing a whiteboard and drawing something completely different. The
    # person holding the whiteboard (the PID) is the same, but the content
    # is entirely new.
    #
    # What changes: registers (zeroed), PC (set to entry point), signal
    #   handlers (cleared), memory.
    # What stays: PID, parent_pid, children, priority, cpu_time.

    def exec(
        self,
        pid: int,
        entry_point: int,
        stack_pointer: int,
        memory_base: int = 0,
        memory_size: int = 0,
    ) -> bool:
        """Exec: replace a process's program with a new one.

        The process keeps its PID, parent, children, and priority, but
        everything else is replaced:
            - Registers are zeroed.
            - PC is set to entry_point.
            - SP is set to stack_pointer.
            - Signal handlers are cleared (the new program doesn't know
              about the old program's handlers).
            - Pending signals are cleared.
            - Memory region is updated if new values are provided.

        Args:
            pid: The PID of the process to exec.
            entry_point: The virtual address where the new program starts.
            stack_pointer: The initial stack pointer for the new program.
            memory_base: New memory base address (0 = keep current).
            memory_size: New memory region size (0 = keep current).

        Returns:
            True if exec succeeded, False if the PID does not exist.

        Example:
            >>> pm = ProcessManager()
            >>> proc = pm.create_process(name="shell")
            >>> pm.exec(proc.pid, entry_point=0x10000, stack_pointer=0x7FFFF000)
            True
            >>> proc.pc
            65536
            >>> proc.registers  # all zeroed
            [0, 0, 0, 0, ..., 0]
        """
        pcb = self._processes.get(pid)
        if pcb is None:
            return False

        # Reset all registers to zero — the new program starts fresh.
        pcb.registers = [0] * 32

        # Set the program counter to the new program's entry point.
        pcb.pc = entry_point

        # Set the stack pointer.
        pcb.sp = stack_pointer

        # Update memory region if new values provided.
        if memory_base != 0:
            pcb.memory_base = memory_base
        if memory_size != 0:
            pcb.memory_size = memory_size

        # Clear signal handlers — the new program has no knowledge of
        # the old program's signal handling setup.
        pcb.signal_handlers.clear()

        # Clear pending signals — the new program starts with a clean
        # signal state.
        pcb.pending_signals.clear()

        return True

    # =========================================================================
    # wait() — Wait for a Child to Exit
    # =========================================================================
    #
    # wait() is how a parent process collects the exit status of its children.
    # Without wait(), terminated child processes become "zombies" — they are
    # dead (no longer running) but their PCBs remain in the process table,
    # consuming resources.
    #
    # The wait() call:
    #   1. Checks if any child is in ZOMBIE state.
    #   2. If yes: reaps the zombie (removes its PCB) and returns its exit code.
    #   3. If no: returns None (in a real kernel, the parent would block).

    def wait(
        self, parent_pid: int, child_pid: int = -1
    ) -> tuple[int, int] | None:
        """Wait for a child to terminate and retrieve its exit code.

        If child_pid is -1, wait for ANY child that is a zombie.
        If child_pid is a specific PID, wait only for that child.

        When a zombie child is found:
            1. The child's exit code is retrieved.
            2. The child is removed from the parent's children list.
            3. The child's PCB is removed from the process table (reaped).

        Args:
            parent_pid: The PID of the waiting parent.
            child_pid: The PID of the specific child to wait for, or -1
                       to wait for any child.

        Returns:
            A tuple (reaped_child_pid, exit_code) if a zombie child was
            found and reaped, or None if no zombie children are available.

        Example:
            >>> pm = ProcessManager()
            >>> parent = pm.create_process(name="shell")
            >>> child_pid, _ = pm.fork(parent.pid)
            >>> pm.exit_process(child_pid, exit_code=42)
            >>> result = pm.wait(parent.pid, child_pid)
            >>> result
            (1, 42)
        """
        parent = self._processes.get(parent_pid)
        if parent is None:
            return None

        # Search the parent's children for a zombie.
        for c_pid in list(parent.children):  # Copy list to allow mutation.
            # If we are waiting for a specific child, skip others.
            if child_pid >= 0 and c_pid != child_pid:
                continue

            child = self._processes.get(c_pid)
            if child is None:
                continue

            if child.state == ProcessState.ZOMBIE:
                # Found a zombie child — reap it!
                exit_code = child.exit_code

                # Remove child from parent's children list.
                parent.children.remove(c_pid)

                # Remove child from the process table entirely.
                del self._processes[c_pid]

                return (c_pid, exit_code)

        # No zombie children found. In a real kernel, the parent would be
        # blocked (state = BLOCKED) until a child exits. For our educational
        # implementation, we return None to indicate "nothing to reap yet."
        return None

    # =========================================================================
    # kill() — Send a Signal to a Process
    # =========================================================================
    #
    # Despite its name, kill() does not necessarily kill a process. It sends
    # a signal, which the process may catch, ignore, or be terminated by.
    # The name comes from its original purpose: sending the default SIGTERM
    # signal, which terminates the process. But you can send any signal.

    def kill(self, pid: int, signal: Signal) -> bool:
        """Send a signal to a process.

        This delegates to the SignalManager, which handles the actual signal
        delivery semantics (immediate action for SIGKILL/SIGSTOP, pending
        queue for everything else).

        Args:
            pid: The PID of the target process.
            signal: The signal to send.

        Returns:
            True if the signal was delivered, False if the PID does not exist.

        Example:
            >>> pm = ProcessManager()
            >>> proc = pm.create_process(name="daemon")
            >>> pm.kill(proc.pid, Signal.SIGTERM)
            True
            >>> proc.pending_signals
            [15]
        """
        process = self._processes.get(pid)
        if process is None:
            return False

        return self._signal_manager.send_signal(process, signal)

    # =========================================================================
    # exit_process() — Terminate a Process
    # =========================================================================
    #
    # When a process calls exit(), or when it reaches the end of its main()
    # function, the kernel:
    #   1. Sets the process state to ZOMBIE.
    #   2. Records the exit code.
    #   3. Reparents any children to PID 0 (init).
    #   4. Sends SIGCHLD to the parent so it knows a child has exited.
    #
    # The process is NOT immediately removed from the process table. It stays
    # as a zombie until the parent calls wait() to retrieve the exit code.
    # This is why you sometimes see zombie processes in `ps` output — their
    # parents haven't waited on them yet.

    def exit_process(self, pid: int, exit_code: int = 0) -> None:
        """Terminate a process.

        Sets the process state to ZOMBIE and records the exit code. The
        process's children are reparented to PID 0 (init process), and
        SIGCHLD is sent to the parent to notify it.

        The PCB remains in the process table until the parent calls wait()
        to reap it.

        Args:
            pid: The PID of the process to terminate.
            exit_code: The exit status code. Convention:
                       0 = success, nonzero = error.

        Example:
            >>> pm = ProcessManager()
            >>> parent = pm.create_process(name="shell")
            >>> child_pid, _ = pm.fork(parent.pid)
            >>> pm.exit_process(child_pid, exit_code=0)
            >>> child = pm.get_process(child_pid)
            >>> child.state
            <ProcessState.ZOMBIE: 4>
        """
        pcb = self._processes.get(pid)
        if pcb is None:
            return

        # Set state to ZOMBIE — the process is dead but not yet reaped.
        pcb.state = ProcessState.ZOMBIE
        pcb.exit_code = exit_code

        # Reparent children to PID 0 (init).
        # In a real system, init periodically calls wait() to reap orphaned
        # zombies. Without reparenting, orphaned children would never be
        # reaped and their PCBs would leak.
        init_process = self._processes.get(0)
        for child_pid_item in pcb.children:
            child = self._processes.get(child_pid_item)
            if child is not None:
                child.parent_pid = 0
                if init_process is not None and pid != 0:
                    init_process.children.append(child_pid_item)

        pcb.children.clear()

        # Send SIGCHLD to the parent so it knows a child has exited.
        # The parent can then call wait() to retrieve the exit code.
        if pcb.parent_pid >= 0:
            parent = self._processes.get(pcb.parent_pid)
            if parent is not None:
                self._signal_manager.send_signal(parent, Signal.SIGCHLD)

    # =========================================================================
    # Query Methods
    # =========================================================================

    def get_process(self, pid: int) -> ProcessControlBlock | None:
        """Get a process by PID.

        Args:
            pid: The PID to look up.

        Returns:
            The ProcessControlBlock, or None if the PID does not exist.
        """
        return self._processes.get(pid)

    def get_children(self, pid: int) -> list[int]:
        """Get the PIDs of all children of a process.

        Args:
            pid: The parent PID.

        Returns:
            List of child PIDs, or empty list if PID not found.
        """
        pcb = self._processes.get(pid)
        if pcb is None:
            return []
        return list(pcb.children)

    def get_parent(self, pid: int) -> int:
        """Get the parent PID of a process.

        Args:
            pid: The PID to query.

        Returns:
            The parent PID, or -1 if the PID does not exist.
        """
        pcb = self._processes.get(pid)
        if pcb is None:
            return -1
        return pcb.parent_pid

    @property
    def signal_manager(self) -> SignalManager:
        """Access the signal manager for direct signal operations."""
        return self._signal_manager

    @property
    def process_count(self) -> int:
        """Return the number of processes in the process table."""
        return len(self._processes)

File: code/packages/go/process-manager/process_manager.go
package processmanager

// =============================================================================
// ProcessManager — Core Process Lifecycle Management
// =============================================================================
//
// This file implements the four fundamental Unix process operations:
//
//   Fork()  — Clone a running process.
//   Exec()  — Replace a process's program with a new one.
//   Wait()  — Wait for a child process to exit and retrieve its exit code.
//   Kill()  — Send a signal to a process.
//
// Together with ExitProcess(), these operations form the complete process
// lifecycle. Every program you have ever run on a Unix system was created
// by fork+exec and cleaned up by exit+wait.
//
// How Your Shell Works
//
// When you type "ls" in a terminal:
//
//  1. The shell calls Fork(). Now there are TWO copies of the shell.
//  2. The child calls Exec("ls"). The child is now the "ls" program.
//  3. "ls" runs, prints files, and calls exit(0). Child becomes a zombie.
//  4. The parent calls Wait(). Gets the child's exit code (0 = success).
//  5. The zombie is reaped (PCB removed). Shell prints the next prompt.

// ProcessManager manages the complete process lifecycle.
//
// It owns the process table (a map from PID to PCB) and provides the system
// calls that user programs invoke to create, manage, and terminate processes.
type ProcessManager struct {
	processes     map[int]*ProcessControlBlock
	nextPID       int
	signalManager *SignalManager
}

// NewProcessManager creates a new ProcessManager with an empty process table.
func NewProcessManager() *ProcessManager {
	result, _ := StartNew[*ProcessManager]("process-manager.NewProcessManager", nil,
		func(op *Operation[*ProcessManager], rf *ResultFactory[*ProcessManager]) *OperationResult[*ProcessManager] {
			return rf.Generate(true, false, &ProcessManager{
				processes:     make(map[int]*ProcessControlBlock),
				nextPID:       0,
				signalManager: NewSignalManager(),
			})
		}).GetResult()
	return result
}

// =============================================================================
// Process Creation
// =============================================================================

// CreateProcess creates a new process with a unique PID.
//
// This is the low-level process creation mechanism. It allocates a fresh PCB,
// assigns a unique PID, and adds it to the process table. Higher-level
// operations like Fork() use this internally.
//
// Parameters:
//   - name: Human-readable name for debugging (e.g., "shell", "ls").
//   - parentPID: PID of the parent process. -1 for the root process.
//   - priority: Scheduling priority, 0 (highest) to 39 (lowest).
//   - memoryBase: Starting address of the process's memory region.
//   - memorySize: Size of the process's memory region in bytes.
func (pm *ProcessManager) CreateProcess(name string, parentPID int, priority int, memoryBase int, memorySize int) *ProcessControlBlock {
	result, _ := StartNew[*ProcessControlBlock]("process-manager.ProcessManager.CreateProcess", nil,
		func(op *Operation[*ProcessControlBlock], rf *ResultFactory[*ProcessControlBlock]) *OperationResult[*ProcessControlBlock] {
			pid := pm.nextPID
			pm.nextPID++

			pcb := NewPCB(pid, name)
			pcb.ParentPID = parentPID
			pcb.Priority = priority
			pcb.MemoryBase = memoryBase
			pcb.MemorySize = memorySize

			pm.processes[pid] = pcb

			// If this process has a parent, add it to the parent's children list.
			if parentPID >= 0 {
				if parent, ok := pm.processes[parentPID]; ok {
					parent.Children = append(parent.Children, pid)
				}
			}

			return rf.Generate(true, false, pcb)
		}).GetResult()
	return result
}

// =============================================================================
// Fork — Clone a Process
// =============================================================================
//
// Fork() is the most unusual system call in computing. It creates a new
// process that is an EXACT COPY of the calling process. Both processes
// resume at the same point in the code, but they receive different return
// values:
//
//   - The parent receives the child's PID (a positive integer).
//   - The child receives 0.
//
// This is how the program knows which copy it is:
//
//   pid := fork()
//   if pid == 0 {
//       fmt.Println("I am the child!")
//   } else {
//       fmt.Printf("I am the parent. My child is PID %d.\n", pid)
//   }

// Fork creates a child process as a copy of the parent.
//
// The child gets:
//   - A new, unique PID.
//   - A copy of the parent's registers (including PC).
//   - The same priority as the parent.
//   - The same signal handlers.
//   - An empty children list and no pending signals.
//   - CPUTime reset to 0.
//
// Returns (childPID, 0, true) on success.
// Returns (0, 0, false) if parentPID does not exist.
//
// In a real kernel, fork() returns childPID to the parent and 0 to the
// child by writing different values into their respective a0 registers.
// Here we return both values so the caller can simulate this behavior.
// forkResult is an internal helper struct for returning multiple values from Fork.
type forkResult struct {
	childPID    int
	childReturn int
	ok          bool
}

func (pm *ProcessManager) Fork(parentPID int) (childPID int, childReturn int, ok bool) {
	res, _ := StartNew[forkResult]("process-manager.ProcessManager.Fork", forkResult{},
		func(op *Operation[forkResult], rf *ResultFactory[forkResult]) *OperationResult[forkResult] {
			parent, exists := pm.processes[parentPID]
			if !exists {
				return rf.Generate(true, false, forkResult{0, 0, false})
			}

			// Allocate a new PID.
			cPID := pm.nextPID
			pm.nextPID++

			// Create the child PCB as a copy of the parent.
			child := NewPCB(cPID, parent.Name)
			child.State = Ready
			child.ParentPID = parentPID
			child.Priority = parent.Priority
			child.MemoryBase = parent.MemoryBase
			child.MemorySize = parent.MemorySize

			// Copy registers — deep copy so modifications are independent.
			child.Registers = parent.Registers // Array copy (value semantics in Go).

			// Copy PC and SP.
			child.PC = parent.PC
			child.SP = parent.SP

			// Copy signal handlers (deep copy the map).
			for sig, handler := range parent.SignalHandlers {
				child.SignalHandlers[sig] = handler
			}

			// Copy signal mask (deep copy the map).
			for sig, blocked := range parent.SignalMask {
				child.SignalMask[sig] = blocked
			}

			// Child starts with: no children, no pending signals, CPUTime = 0.
			// These are already set by NewPCB.

			// Add child to process table.
			pm.processes[cPID] = child

			// Update parent's children list.
			parent.Children = append(parent.Children, cPID)

			return rf.Generate(true, false, forkResult{cPID, 0, true})
		}).GetResult()
	return res.childPID, res.childReturn, res.ok
}

// =============================================================================
// Exec — Replace Process Image
// =============================================================================
//
// Exec() throws away the current program and loads a new one. It is like
// erasing a whiteboard and drawing something completely different. The
// person holding the whiteboard (the PID) is the same, but the content
// is entirely new.
//
// What changes: registers (zeroed), PC (entry point), signal handlers (cleared).
// What stays: PID, ParentPID, Children, Priority, CPUTime.

// Exec replaces a process's program with a new one.
//
// Registers are zeroed, PC is set to entryPoint, SP is set to stackPointer,
// signal handlers and pending signals are cleared. PID, parent, children,
// and priority are preserved.
//
// Returns true on success, false if the PID does not exist.
func (pm *ProcessManager) Exec(pid int, entryPoint int, stackPointer int, memoryBase int, memorySize int) bool {
	result, _ := StartNew[bool]("process-manager.ProcessManager.Exec", false,
		func(op *Operation[bool], rf *ResultFactory[bool]) *OperationResult[bool] {
			pcb, exists := pm.processes[pid]
			if !exists {
				return rf.Generate(true, false, false)
			}

			// Reset all registers to zero — fresh start for the new program.
			pcb.Registers = [NumRegisters]int{}

			// Set the program counter to the new entry point.
			pcb.PC = entryPoint

			// Set the stack pointer.
			pcb.SP = stackPointer

			// Update memory region if new values provided.
			if memoryBase != 0 {
				pcb.MemoryBase = memoryBase
			}
			if memorySize != 0 {
				pcb.MemorySize = memorySize
			}

			// Clear signal handlers — the new program has no knowledge of the
			// old program's signal handling setup.
			pcb.SignalHandlers = make(map[int]int)

			// Clear pending signals.
			pcb.PendingSignals = make([]int, 0)

			return rf.Generate(true, false, true)
		}).GetResult()
	return result
}

// =============================================================================
// Wait — Wait for a Child to Exit
// =============================================================================
//
// Wait() is how a parent process collects the exit status of its children.
// Without Wait(), terminated children become "zombies" — dead but their
// PCBs remain in the process table.

// Wait waits for a child to terminate and retrieves its exit code.
//
// If childPID is -1, waits for ANY child that is a zombie.
// If childPID is a specific PID, waits only for that child.
//
// Returns (reapedPID, exitCode, true) if a zombie child was found and reaped.
// Returns (0, 0, false) if no zombie children are available or parent doesn't
// exist.
// waitResult is an internal helper struct for returning multiple values from Wait.
type waitResult struct {
	reapedPID int
	exitCode  int
	ok        bool
}

func (pm *ProcessManager) Wait(parentPID int, childPID int) (reapedPID int, exitCode int, ok bool) {
	res, _ := StartNew[waitResult]("process-manager.ProcessManager.Wait", waitResult{},
		func(op *Operation[waitResult], rf *ResultFactory[waitResult]) *OperationResult[waitResult] {
			parent, exists := pm.processes[parentPID]
			if !exists {
				return rf.Generate(true, false, waitResult{0, 0, false})
			}

			// Search the parent's children for a zombie.
			for i, cPID := range parent.Children {
				// If waiting for a specific child, skip others.
				if childPID >= 0 && cPID != childPID {
					continue
				}

				child, childExists := pm.processes[cPID]
				if !childExists {
					continue
				}

				if child.State == Zombie {
					// Found a zombie child — reap it!
					code := child.ExitCode

					// Remove child from parent's children list.
					parent.Children = append(parent.Children[:i], parent.Children[i+1:]...)

					// Remove child from process table entirely.
					delete(pm.processes, cPID)

					return rf.Generate(true, false, waitResult{cPID, code, true})
				}
			}

			// No zombie children found.
			return rf.Generate(true, false, waitResult{0, 0, false})
		}).GetResult()
	return res.reapedPID, res.exitCode, res.ok
}

// =============================================================================
// Kill — Send a Signal to a Process
// =============================================================================
//
// Despite its name, Kill() does not necessarily kill a process. It sends
// a signal, which the process may catch, ignore, or be terminated by.

// Kill sends a signal to a process.
//
// Returns true if the signal was delivered, false if the PID does not exist.
func (pm *ProcessManager) Kill(pid int, signal int) bool {
	result, _ := StartNew[bool]("process-manager.ProcessManager.Kill", false,
		func(op *Operation[bool], rf *ResultFactory[bool]) *OperationResult[bool] {
			process, exists := pm.processes[pid]
			if !exists {
				return rf.Generate(true, false, false)
			}
			return rf.Generate(true, false, pm.signalManager.SendSignal(process, signal))
		}).GetResult()
	return result
}

// =============================================================================
// ExitProcess — Terminate a Process
// =============================================================================
//
// When a process exits:
//  1. State is set to Zombie.
//  2. Exit code is recorded.
//  3. Children are reparented to PID 0 (init).
//  4. SIGCHLD is sent to the parent.

// ExitProcess terminates a process.
//
// Sets state to Zombie, records the exit code, reparents children to PID 0,
// and sends SIGCHLD to the parent.
func (pm *ProcessManager) ExitProcess(pid int, exitCode int) {
	_, _ = StartNew[struct{}]("process-manager.ProcessManager.ExitProcess", struct{}{},
		func(op *Operation[struct{}], rf *ResultFactory[struct{}]) *OperationResult[struct{}] {
			pcb, exists := pm.processes[pid]
			if !exists {
				return rf.Generate(true, false, struct{}{})
			}

			pcb.State = Zombie
			pcb.ExitCode = exitCode

			// Reparent children to PID 0 (init).
			initProcess, initExists := pm.processes[0]
			for _, childPID := range pcb.Children {
				child, childExists := pm.processes[childPID]
				if childExists {
					child.ParentPID = 0
					if initExists && pid != 0 {
						initProcess.Children = append(initProcess.Children, childPID)
					}
				}
			}
			pcb.Children = make([]int, 0)

			// Send SIGCHLD to the parent.
			if pcb.ParentPID >= 0 {
				if parent, parentExists := pm.processes[pcb.ParentPID]; parentExists {
					pm.signalManager.SendSignal(parent, SIGCHLD)
				}
			}
			return rf.Generate(true, false, struct{}{})
		}).GetResult()
}

// =============================================================================
// Query Methods
// =============================================================================

// GetProcess returns the PCB for the given PID, or nil if not found.
func (pm *ProcessManager) GetProcess(pid int) *ProcessControlBlock {
	result, _ := StartNew[*ProcessControlBlock]("process-manager.ProcessManager.GetProcess", nil,
		func(op *Operation[*ProcessControlBlock], rf *ResultFactory[*ProcessControlBlock]) *OperationResult[*ProcessControlBlock] {
			return rf.Generate(true, false, pm.processes[pid])
		}).GetResult()
	return result
}

// GetChildren returns the PIDs of all children of a process.
func (pm *ProcessManager) GetChildren(pid int) []int {
	result, _ := StartNew[[]int]("process-manager.ProcessManager.GetChildren", nil,
		func(op *Operation[[]int], rf *ResultFactory[[]int]) *OperationResult[[]int] {
			pcb, exists := pm.processes[pid]
			if !exists {
				return rf.Generate(true, false, nil)
			}
			children := make([]int, len(pcb.Children))
			copy(children, pcb.Children)
			return rf.Generate(true, false, children)
		}).GetResult()
	return result
}

// GetParent returns the parent PID of a process, or -1 if not found.
func (pm *ProcessManager) GetParent(pid int) int {
	result, _ := StartNew[int]("process-manager.ProcessManager.GetParent", -1,
		func(op *Operation[int], rf *ResultFactory[int]) *OperationResult[int] {
			pcb, exists := pm.processes[pid]
			if !exists {
				return rf.Generate(true, false, -1)
			}
			return rf.Generate(true, false, pcb.ParentPID)
		}).GetResult()
	return result
}

// SignalManager returns the process manager's signal manager.
func (pm *ProcessManager) SignalManager() *SignalManager {
	result, _ := StartNew[*SignalManager]("process-manager.ProcessManager.SignalManager", nil,
		func(op *Operation[*SignalManager], rf *ResultFactory[*SignalManager]) *OperationResult[*SignalManager] {
			return rf.Generate(true, false, pm.signalManager)
		}).GetResult()
	return result
}

// ProcessCount returns the number of processes in the process table.
func (pm *ProcessManager) ProcessCount() int {
	result, _ := StartNew[int]("process-manager.ProcessManager.ProcessCount", 0,
		func(op *Operation[int], rf *ResultFactory[int]) *OperationResult[int] {
			return rf.Generate(true, false, len(pm.processes))
		}).GetResult()
	return result
}
